lib/cli/kit/table_formatter.rb

```diff
@@ -0,0 +1,161 @@
+    # which must support a method for each column. In our examples, we use a
+    # `Struct` to implement these, but if one or more of your columns is
+    # expensive to generate and unlikely to be printed, this provides you an
+    # easy way to lazy-generate it.
+    #
+    class TableFormatter
+      # Instantiate a new `TableFormatter`, with a list of supported columns.
+      # If `default` is provided, it will determine which columns are printed
+      # by default, e.g.:
+      #
+      #     TableFormatter.new(:pid, :ppid, :cmd, default: [:pid, :ppid])
+      #
+      # In this case, only the pid and ppid columns will be printed by default,
+      # but cmd will be available for selection with -o.
+      def initialize(*columns, default: nil)
+        @columns       = columns
+        @rows          = []
+        @print_columns = default || @columns.dup
+        @print_header  = true
+      end
+
+      # add the -o and -H options to an `OptionParser` instance. Normally, you
+      # will only need to call `parse_options`, but `add_options` is useful if
+      # you also want to support other options that `TableFormatter` doesn't
+      # know or care about.
+      def add_options(opt_parser)
+        opt_parser.on(
+          '-oCOLUMNS', '--output=COLUMNS',
+          'print only certain comma-separated columns, e.g. "pid,ppid"',
+        ) do |v|
+          cols = v.split(',').map(&:to_sym)
+          extra_cols = cols - @columns
+          if extra_cols.any?
+            raise(Dev::Abort, "invalid column: #{extra_cols.join(', ')}")
+          end
+          @print_columns = cols
+        end
+
+        opt_parser.on('-H', '--no-header', "don't print the header row") do
+          @print_header = false
+        end
+      end
+
+      # Add a row to the table. The object passed must support each column as a
+      # method.
+      def <<(obj)
+        @rows << obj
+      end
+
+      # Parse command-line options -o and -H, storing the result internally to the TableFormatter.
+      def parse_options(*argv)
+        opt_parser = OptionParser.new { |opts| add_options(opts) }
+        opt_parser.parse!(argv)
+        argv
+      end
+
+      # Having `<<`'d all the data, render the formatted table to output
+      # (`to`). This is written in a slightly over-factored way in case it's
+      # useful to override any part.
+      def render(to: STDOUT, columns: @print_columns, header: @print_header)
+        out = build_header(columns: columns, header: header)
+        @rows.each { |row| out << build_row(row, columns: columns) }
+        widths = calculate_widths(out)
+        fmtstr = build_fmtstr(widths)
+        write_output(fmtstr, out, to)
+      end
+
+      private
+
+      def build_header(columns:, header:)
+        if header
+          [columns.map { |col| col.to_s.upcase }]
+        else
+          []
+        end
+      end
+
+      def build_row(row_obj, columns:)
+        columns.map { |col| row_obj.send(col).to_s }
```

Comment:
A hash might be less cumbersome to use than a Struct sometimes. Wanna catch `NoMethodError` and fallback to `row_obj[col]`?